File: client/src/api/threads.ts
import { apiRequest } from "../services/api";
import type { Message, MessagePage } from "./channels";

export interface CreateThreadResponse {
  thread_id: string;
  parent_message_id: string;
  channel_id: string;
  created_at: string;
  reply_count: number;
  follower_count: number;
}

export interface Thread {
  id: string;
  channel_id: string;
  parent_message_id: string;
  creator_id: string;
  created_at: string;
  last_reply_at: string | null;
  reply_count: number;
  follower_count: number;
  is_following: boolean;
}

export async function createThread(
  address: string,
  token: string,
  channelId: string,
  messageId: string,
): Promise<CreateThreadResponse> {
  return apiRequest<CreateThreadResponse>(
    address,
    `/api/v1/channels/${channelId}/messages/${messageId}/threads`,
    { method: "POST", token },
  );
}

export async function getThread(
  address: string,
  token: string,
  threadId: string,
): Promise<Thread> {
  return apiRequest<Thread>(address, `/api/v1/threads/${threadId}`, { token });
}

export async function listThreadMessages(
  address: string,
  token: string,
  threadId: string,
  before?: string,
  after?: string,
  limit?: number,
): Promise<MessagePage> {
  const params = new URLSearchParams();
  if (before) params.set("before", before);
  if (after) params.set("after", after);
  if (limit) params.set("limit", limit.toString());

  const query = params.toString();
  const path = `/api/v1/threads/${threadId}/messages${query ? `?${query}` : ""}`;
  return apiRequest<MessagePage>(address, path, { token });
}

export async function createThreadMessage(
  address: string,
  token: string,
  threadId: string,
  content: string,
  attachmentIds: string[] = [],
): Promise<Message> {
  return apiRequest<Message>(address, `/api/v1/threads/${threadId}/messages`, {
    method: "POST",
    token,
    body: JSON.stringify({ content, attachment_ids: attachmentIds }),
  });
}

export async function followThread(
  address: string,
  token: string,
  threadId: string,
): Promise<void> {
  return apiRequest<void>(address, `/api/v1/threads/${threadId}/follow`, {
    method: "PUT",
    token,
  });
}

export async function unfollowThread(
  address: string,
  token: string,
  threadId: string,
): Promise<void> {
  return apiRequest<void>(address, `/api/v1/threads/${threadId}/follow`, {
    method: "DELETE",
    token,
  });
}

export async function markThreadRead(
  address: string,
  token: string,
  threadId: string,
): Promise<void> {
  return apiRequest<void>(address, `/api/v1/threads/${threadId}/read`, {
    method: "PUT",
    token,
  });
}

File: sdk/rust/src/rest/threads.rs
use reqwest::Method;

use crate::error::Result;
use crate::rest::models::{
    CreateMessageRequest, CreateThreadResponse, ListMessagesQuery, Message, MessagePage, Thread,
};
use crate::rest::Client;

pub struct ThreadsApi<'c> {
    client: &'c Client,
}

impl<'c> ThreadsApi<'c> {
    pub(crate) fn new(client: &'c Client) -> Self {
        Self { client }
    }

    pub async fn create(
        &self,
        channel_id: &str,
        message_id: &str,
    ) -> Result<CreateThreadResponse> {
        self.client
            .post_empty(&format!(
                "/api/v1/channels/{channel_id}/messages/{message_id}/threads"
            ))
            .await
    }

    pub async fn get(&self, thread_id: &str) -> Result<Thread> {
        self.client
            .get(&format!("/api/v1/threads/{thread_id}"))
            .await
    }

    pub async fn list_messages(
        &self,
        thread_id: &str,
        query: &ListMessagesQuery,
    ) -> Result<MessagePage> {
        let path = format!("/api/v1/threads/{thread_id}/messages");
        let req = self.client.request(Method::GET, &path)?.query(query);
        let resp = req.send().await?;
        super::decode_response(resp).await
    }

    pub async fn post_message(
        &self,
        thread_id: &str,
        req: &CreateMessageRequest,
    ) -> Result<Message> {
        self.client
            .post(&format!("/api/v1/threads/{thread_id}/messages"), req)
            .await
    }

    pub async fn follow(&self, thread_id: &str) -> Result<()> {
        self.client
            .put_empty(&format!("/api/v1/threads/{thread_id}/follow"))
            .await
    }

    pub async fn unfollow(&self, thread_id: &str) -> Result<()> {
        self.client
            .delete(&format!("/api/v1/threads/{thread_id}/follow"))
            .await
    }

    pub async fn mark_read(&self, thread_id: &str) -> Result<()> {
        self.client
            .put_empty(&format!("/api/v1/threads/{thread_id}/read"))
            .await
    }
}
